### PR TITLE
feat: Support msg to create policy by cross chain

### PR DESCRIPTION
**Pull request by constwz**
### Description

1. Cross-chain policy creation supports msg
2. Remove cross-chain operation restrictions https://docs.google.com/spreadsheets/d/1CH2W0kNgQXwQIhQybdcxaeKm4QtEnse_3tuMd0u6zAQ/edit?gid=0#gid=0

### CODE REVIEW

**Review by cosinlink — APPROVED** [review]

**Review by yutianwu — APPROVED** [review]

File: x/storage/keeper/cross_app_permission.go [hunk 1]
```diff
@@ -146,16 +147,88 @@ func (app *PermissionApp) handleCreatePolicySynPackage(ctx sdk.Context, createPo
...
 	err = policy.Unmarshal(createPolicyPackage.Data)
 	if err != nil {
-		return sdk.ExecuteResult{
-			Payload: types.CreatePolicyAckPackage{
-				Status:    types.StatusFail,
-				Creator:   createPolicyPackage.Operator,
-				ExtraData: createPolicyPackage.ExtraData,
-			}.MustSerialize(),
-			Err: err,
+		var msgPutPolicy types.MsgPutPolicy
```
Comment (unclezoro) [inline x/storage/keeper/cross_app_permission.go:153]: Isn't this is a hardfork logic?

File: x/storage/keeper/cross_app_permission.go [hunk 2]
```diff
@@ -146,16 +147,88 @@ func (app *PermissionApp) handleCreatePolicySynPackage(ctx sdk.Context, createPo
...
+			return sdk.ExecuteResult{
+				Payload: types.CreatePolicyAckPackage{
+					Status:    types.StatusFail,
+					Creator:   createPolicyPackage.Operator,
+					ExtraData: createPolicyPackage.ExtraData,
+				}.MustSerialize(),
+				Err: err,
+			}
+		}
+		policy.ResourceId = resID
```
Comment (unclezoro) [inline x/storage/keeper/cross_app_permission.go:231]: To me, this feature is to support resource name, so this assign is the key point. 
But before this, a lot of code is added to do the verification, what is the motivation? Is the previous implementation missed to do verification?

File: x/storage/keeper/cross_app_permission.go (hunk 1 above)
Comment (unclezoro) [inline x/storage/keeper/cross_app_permission.go:153]: I think define another type `MsgPutPolicy` is not friendly to the developer on BSC side, they have to distinguage `permtypes.Policy` and `types.MsgPutPolicy`, that is quite confusing. Please extend the `permtypes.Policy`

File: x/storage/keeper/cross_app_permission.go [hunk 3]
```diff
@@ -243,3 +316,45 @@ func (app *PermissionApp) getResourceOwner(ctx sdk.Context, policy *permtypes.Po
 	}
 	return resOwner, nil
 }
+
+func (app *PermissionApp) getResourceOwnerAndIdFromGRN(ctx sdk.Context, grn types2.GRN) (resOwner sdk.AccAddress, resID math.Uint, err error) {
```
Comment (unclezoro) [inline x/storage/keeper/cross_app_permission.go:320]: why copy so many code, please use the `getResourceOwnerAndIdFromGRN` in the permission.go

**Review by cosinlink — DISMISSED** [review]

File: x/storage/keeper/cross_app_permission.go (hunk 3 above)
Comment (constwz) [reply x/storage/keeper/cross_app_permission.go:320]: reject, getResourceOwnerAndIdFromGRN in the permission.go is a private function belong to Keeper

File: x/storage/keeper/keeper.go [hunk 4]
```diff
@@ -221,7 +221,7 @@ func (k Keeper) DeleteBucket(ctx sdk.Context, operator sdk.AccAddress, bucketNam
 	if !found {
 		return types.ErrNoSuchBucket
 	}
-	if bucketInfo.SourceType != opts.SourceType {
+	if bucketInfo.SourceType != types.SOURCE_TYPE_ORIGIN && bucketInfo.SourceType != opts.SourceType {
```
Comment (unclezoro) [inline x/storage/keeper/keeper.go:224]: please add hardfork logic here.

File: x/storage/keeper/cross_app_permission.go (hunk 2 above)
Comment (constwz) [reply x/storage/keeper/cross_app_permission.go]: accept,fixed

File: x/storage/keeper/cross_app_permission.go (hunk 1 above)
Comment (constwz) [reply x/storage/keeper/cross_app_permission.go]: accept,fixed

File: x/storage/keeper/cross_app_permission.go (hunk 3 above)
Comment (constwz) [reply x/storage/keeper/cross_app_permission.go]: accept,fixed

**Review by unclezoro — DISMISSED** [review]

**Review by cosinlink — APPROVED** [review]

**Review by unclezoro — APPROVED** [review]

**Review by cosinlink — APPROVED** [review]

**Pull request merged**